zaza/utilities/openstack_provider.py

```diff
@@ -0,0 +1,305 @@
+        domain = os.environ.get('OS_DOMAIN_NAME')
+        auth_settings = {
+            'OS_AUTH_URL': os.environ.get('OS_AUTH_URL'),
+            'OS_USERNAME': os.environ.get('OS_USERNAME'),
+            'OS_PASSWORD': os.environ.get('OS_PASSWORD'),
+            'OS_REGION_NAME': os.environ.get('OS_REGION_NAME'),
+            'API_VERSION': 3,
+        }
+        if domain:
+            auth_settings['OS_DOMAIN_NAME'] = domain
+        else:
+            auth_settings['OS_USER_DOMAIN_NAME'] = (
+                os.environ.get('OS_USER_DOMAIN_NAME'))
+            auth_settings['OS_PROJECT_NAME'] = (
+                os.environ.get('OS_PROJECT_NAME'))
+            auth_settings['OS_PROJECT_DOMAIN_NAME'] = (
+                os.environ.get('OS_PROJECT_DOMAIN_NAME'))
+            os_project_id = os.environ.get('OS_PROJECT_ID')
+            if os_project_id is not None:
+                auth_settings['OS_PROJECT_ID'] = os_project_id
+
+    _os_cacert = os.environ.get('OS_CACERT')
+    if _os_cacert:
+        auth_settings.update({'OS_CACERT': _os_cacert})
+
+    # Validate settings
+    for key, settings in list(auth_settings.items()):
+        if settings is None:
+            logging.error('Missing OS authentication setting: {}'
+                          ''.format(key))
+            # raise exceptions.MissingOSAthenticationException(
+            raise MissingOSAthenticationException(
+                'One or more OpenStack authentication variables could '
+                'be found in the environment. Please export the OS_* '
+                'settings into the environment.')
+
+    return auth_settings
+
+
+# Nova utilities
+def get_nova_session_client(session, version=2):
+    """Return novaclient authenticated by keystone session.
+
+    :param session: Keystone session object
+    :type session: keystoneauth1.session.Session object
+    :param version: Version of client to request.
+    :type version: float
+    :returns: Authenticated novaclient
+    :rtype: novaclient.Client object
+    """
+    return novaclient_client.Client(version, session=session)
+
+
+# Manage resources
+def delete_resource(resource, resource_id, msg="resource"):
+    """Delete an openstack resource.
+
+    Delete an openstack resource, such as one instance, keypair,
+    image, volume, stack, etc., and confirm deletion within max wait time.
+
+    :param resource: pointer to os resource type, ex:glance_client.images
+    :type resource: str
+    :param resource_id: unique name or id for the openstack resource
+    :type resource_id: str
+    :param msg: text to identify purpose in logging
+    :type msg: str
+    """
+    logging.debug('Deleting OpenStack resource '
+                  '{} ({})'.format(resource_id, msg))
+    resource.delete(resource_id)
+    resource_removed(resource, resource_id, msg)
+
+
+def _resource_removed(resource, resource_id, msg="resource"):
+    """Wait for an openstack resource to no longer be present.
```

Comment:
This function isn't waiting but immediately asserting

Reply:
Actually, it's also poorly named; I'll update this bit.